### ISSUE TITLE
Use LangChain for LLM/AI application code?

### ISSUE THREAD

**Issue by jurmy24** [Tanzania-AI-Community/twiga#70]
Right now we're using Together AI's and OpenAI's SDKs for development but it may be neater and better for future developments to use LangChain. 

Investigate this and respond in comment whether to go through with it or not for v0.2.0.

**Comment by jurmy24** [comment]
Step 1. Understand the file llm_service.py and llm_utils.py
Step 2. Look at https://python.langchain.com/docs/integrations/llms/together/ and https://python.langchain.com/docs/integrations/llms/openai/
Step 3. Replace the existing way to call these services using the langchain sdk
Step 4. Look at further files that calls the LLM providers (embedder.py)

After doing this you can remove the together-ai and openai installations in the .venv by running uv remove ___.

**Comment by smartwhale8** [comment]
## Overview of functionality
### llm_service.py
app/services/llm_service.py defines `LLMClient` class. 
It is responsible for handling interactions with LLMs for the following purposes:
1. Message processing
2. Handling tool calls
3. Generating responses

Key components include:

- **MessageProcessor**: manages message batching and processing for a single user.
- **LLMClient**: handles the main logic for interacting with the LLMs, including:
    - retrieving or creating a `MessageProcessor` for a user.
    - cleaning up message processors if they are empty and unlocked
    - parse response text to extract tool calls if present
    - sending notifications to users when a tool call is made
    - processing tool calls and returning new tool response messages
    - generating a response by handling message batching and tool calls
    - formatting messages for the API, removing duplicates between new messages and database history.

### llm_utils.py
app/utils/llm_utils.py provides utility functions for interacting with the LLM, including token counting and making API requests. 

Key components include:
- **num_tokens_from_string**: returns the number of OpenAI-equivalent tokens in a text-string
- **num_tokens_from_messages**: returns the number of tokens used by a list of messages in the format sent to the OpenAI API
- **async_llm_request**: makes a request to the LLM API with exponential backoff retry logic 

## Static analysis for switching from OpenAI to Langchain
In order to switch to LangChain API, we would need to make the following changes:

### llm_utils.py
- In order to change from OpenAI's `ChatCompletion` API, we would need LangChain's [OpenAI Integration](https://python.langchain.com/api_reference/openai/index.html), specifically:
 LLM Integration `OpenAI` chat model, `ChatOpenAI` [message types](https://python.langchain.com/api_reference/core/messages.html) `HumanMessage`, `SystemMessage`, `AIMessage`, and `FunctionMessage`.
- Replacing OpenAI Client initialization with LangChain initialization (using `ChatOpenAI(model_name, openai_api_key, openai_api_base)`)
- Replacing `async_llm_request() -> ChatCompletion` with a version which uses LangChain. This would involve reconstructing the entire message structure that is sent to the LLM, specifically, converting message format from OpenAI to LangChain: append message types SystemMessage (for system), HumanMessage (for user), AIMessage (for assistant), FunctionMessage (for tool, or function).
- Handling tool configurations for LangChain by converting OpenAI tool format to [LangChain's tool format](https://python.langchain.com/docs/how_to/custom_tools/).

### llm_service.py
- Import required types - HumanMessage, SystemMessage, AIMessage, FunctionMessage & StructuredTool
- The return message format in `_format_messages` need not change. It can continue to return the messages in OpenAI-compatible format because the conversion from OpenAI format to LangChain format will happen later in the `async_llm_request` function in llm_utils.py (where we construct each piece of the LangChain message using SystemMessage, HumanMessage etc.
- Change Tool processing logic in `generate_response` method (where we are calling LLMs with tool enabled `async_llm_request()`) because the LangChain tools need to be called differently. To handle LangChain's tool format, we might need to create a function that coverts the tool metadata to LangChain format.
- We need to update how `tool_calls` are handled. Since LangChain structures this data differently, we need to extract it from the response format.

### embedder.py
- LangChain provides [`OpenAIEmbeddings`](https://python.langchain.com/api_reference/openai/embeddings/langchain_openai.embeddings.base.OpenAIEmbeddings.html#langchain_openai.embeddings.base.OpenAIEmbeddings)
- We need to change the way the client is initialized with LangChain version:
  ```python
  from 
      client = OpenAI(api_key)
  to 
      embeddings_client = OpenAIEmbeddings(
                  llm_settings.embedding_model, 
                  openai_api_key, 
                  openai_api_base
              )
  ```
- In `get_embedding()` function, we need to change to LangChain's version, which is actually much more concise than OpenAI's way of calling:
  ```python
    embedding = embeddings_client.embed_query(text)
  ```
- In `get_embeddings()` we do the similar as above, but this time calling `embed_documents()` for creating embedding for multiple texts:
  ```python
      embeddings = embeddings_client.embed_documents(texts)
  ```

To summarize, the amount of changes might not be much, but would require extensive testing.
